src/OFS/Traversable.py

```diff
@@ -168,15 +168,20 @@ def unrestrictedTraverse(self, path, default=_marker, restricted=False):
         validated with the security machinery. Usually invoked using
         restrictedTraverse().
         """
+        if path:
+            if isinstance(path, str):
+                path = path.split('/')
+            else:
+                path = list(path)
+                for part in path:
+                    if not isinstance(part, str):
+                        raise TypeError(
+                            "path should be a string or an iterable of strings"
+                        )
+
```

Comment:
I suggest you retain the original order and drop this code below the `if not path` test.

Reply:
Instead I would leave the patch as it is because path changes under `if path` and so it make sense to reevaluate it after the first if.

Reply:
You're right that the path may change, but I see no possibility in that code that it can change from a truth value to a non-truth value. That is not possible.

Reply:
Well, we would need an object that is truish but casted to a list returns an empty list...
```
ipdb> self.context.invokeFactory("Folder", "folder")
'folder'
ipdb> bool(self.context["folder"])
True
ipdb> !list(self.context["folder"])
[]
```

Not that I am encouraging people to use a folder as an argument to `unrestrictedTraverse`, but you know:

> I've seen things you people wouldn't believe...

So in my opinion leaving the patch as it is is better :)

Reply:
@dataflake did I make you change your mind?

Reply:
I don't think the argument that people might throw garbage at this method is valid. The method has very clear documentation about what it accepts. Everything else simply deserves to lose. You cannot guard against everything, and attempting to do that is a non-goal.

The original intent of your changes was to tell people if they sent in the wrong kind of string. But you conflate that with changing the control flow, which was very clear and obvious before your changes. Please keep it that way.